rfcs/90-re-support-initialization-via-connection-string.md

```diff
@@ -0,0 +1,152 @@
+---
+author: xxchan <xxchan22f@gmail.com>
+status: draft
+updated_at: 2021-06-09
+deprecates:
+  - 13-remove-config-string.md
+---
+
+# GSP-90: Re-support Initialization Via Connection String
+
+## Background
+
+We had config strings like `s3://bucket_name?access_key=xxxx&secret_key=xxxx` before. This is introduced in [GSP-3] but deprecated by [GSP-13].
+
+There are two reasons why config strings are removed, as mentioned in [GSP-13]:
+
+1. Usability: If an application uses [go-storage]'s config string, then it either exposes the format of config string to end user directly (additional burden for end users) or it has to manipulate its config into our format (additional burden for application developers). And if the application wants to combine its other config into a config string, this is also inconvenient.
+2. Implementation: We didn't find a good way to parse pairs from strings easily and safely.
+
+So we let the users of [go-storage] construct pairs directly using `WithXxx()`. 
+
+But passing string config is indeed more convenient, if not much. And actually parsing string into pairs is not that difficult. We can use a registry mechanism like [GSP-48]'s implementation.
+
+## Proposal
+
+So I propose to support service init from connection string (We rename config string to connection string since this name is more common):
+
+We add the following APIs:
+```go
+func NewServicerFromString(connStr string) (types.Servicer, error) {}
+func NewStoragerFromString(connStr string) (types.Storager, error) {}
+```
+
+### Format
+
+The format of the connection string is (optional parts marked by squared brackets):
+
+`<type>://[<name>][<work_dir>][?key1=value1&...&keyN=valueN]`
+
+- name: storage name, e.g., bucket name. MUST NOT contain /
+- work_dir: For object storage, it is prefix; for fs, it is directory path. MUST start with / for every storage services.
+- At least one of name and work_dir MUST be present.
+- For the `key=value` pairs, if `=value` is missing, we just ignore the pair. But `key=` means a pair with a blank value.
+
+So a valid connection string could be:
+
+- `s3://bucket_name`
+- `s3://bucket_name/prefix`
+- `s3://bucket_name/prefix?credential=hmac:xxxx:xxxx&endpoint=http://bucket_name.s3-website-Region.amazonaws.com`
```

Comment:
Maybe s3 endpoint should not contain `bucket_name`? As described in [Amazon Simple Storage Service endpoints and quotas](https://docs.aws.amazon.com/general/latest/gr/s3.html):

```
s3.us-east-1.amazonaws.com
s3-fips.us-east-1.amazonaws.com
s3.amazonaws.com
s3.dualstack.us-east-1.amazonaws.com**
```